### PR TITLE
[wip] support service account key creation

### PR DESCRIPTION
**Pull request by honnix**
start from here?

### CODE REVIEW

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 1]
```diff
@@ -126,37 +136,120 @@
...
+        throw new InvalidExecutionException(
+            "Referenced secret '" + specSecret.name() + "' was not found");
+      } else {
+        LOG.info("[AUDIT] Workflow {} refers to secret {}",
+                 workflowInstance.workflowId(), specSecret.name());
+      }
+    }
+  }
+
+  private static String buildSecretName(String serviceAccount) {
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: include sanitized prefix substrings of the component, workflow and service account names in the secret name to ease debugging?

e.g. `styx-wf-sa-keys.foo-comp.bar-workfl.baz-acct-d34db33fc4f3d00d` ?

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 2]
```diff
@@ -126,37 +136,120 @@
...
+      spec = spec.addNewVolume()
+          .withName(STYX_WORKFLOW_SA_SECRET_NAME)
+          .withNewSecret()
+          .withSecretName(buildSecretName(runSpec.serviceAccount().get()))
+          .endSecret()
+          .endVolume();
+      container = container
+          .addToVolumeMounts(new VolumeMount(STYX_MANAGED_WORKFLOW_SA_SECRET_MOUNT_PATH,
+                                             STYX_WORKFLOW_SA_SECRET_NAME, true))
+          // TODO: do we need set this env as default value?
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: 👍 

This way we can migrate to network (metadata server) based SA credentials in the future without the user having to care.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 3]
```diff
@@ -126,37 +136,120 @@
...
-      });
-
       final Pod pod = client.pods().create(createPod(workflowInstance, runSpec));
       return pod.getMetadata().getName();
     } catch (KubernetesClientException kce) {
       throw new IOException("Failed to create Kubernetes pod", kce);
     }
   }
 
+  private void ensureSecret(WorkflowInstance workflowInstance, RunSpec runSpec) {
```
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Perhaps SA key creation could be factored into a separate class?

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 1 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Do we want store replicated secrets? Now we have one `service_account` mapped to one `kube secret`. By prefixing component and workflow, we will end up having many secrets having keys pointing to the same SA.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 3 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: 👍 You are on top of that already.

File: styx-api-service/src/test/java/com/spotify/styx/api/deprecated/WorkflowResourceTest.java [hunk 4]
```diff
@@ -89,7 +89,8 @@ public WorkflowResourceTest(Api.Version version) {
 
   private static final WorkflowConfiguration WORKFLOW_CONFIGURATION =
       WorkflowConfiguration
-          .create("bar", Schedule.DAYS, empty(), empty(), empty(), empty(), empty(), emptyList());
+          .create("bar", Schedule.DAYS, empty(), empty(), empty(), empty(), empty(), empty(),
```
Comment (bergman) [inline styx-api-service/src/test/java/com/spotify/styx/api/deprecated/WorkflowResourceTest.java]: I vote for builders.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 5]
```diff
@@ -109,54 +123,156 @@
...
+                                    .withName(secretName)
+                                    .withAnnotations(annotations)
+                                    .endMetadata()
+                                    .withData(keys)
+                                    .build());
+        LOG.info("[AUDIT] Secret {} created for service account {}", secretName, serviceAccount);
+      }
+    }
+
+    if (runSpec.secret().isPresent()) {
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: @bergman @fabriziodemaria @danielnorberg let's continue here. do we want them to co-eixst.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 2 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: There is however one issue. This env will overwrite what's put in `Dockerfile` by user when building image.

File: styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesDockerRunnerTest.java [hunk 6]
```diff
@@ -140,6 +173,89 @@ public void tearDown() throws Exception {
     kdr.close();
   }
 
+  // TODO
```
Comment (danielnorberg) [inline styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesDockerRunnerTest.java]: WIP! Wasn't immediately obvious to me how to fix the mocks in these tests so I commented them out.

Comment (danielnorberg) [inline styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesDockerRunnerTest.java]: Temporarily, will fix later unless you do it first.

**Comment by bergman** [comment]
👍

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 7]
```diff
@@ -111,106 +125,139 @@
...
+    if (runSpec.serviceAccount().isPresent()) {
+      spec.addNewVolume()
+          .withName(STYX_WORKFLOW_SA_SECRET_NAME)
+          .withNewSecret()
+          .withSecretName(buildSecretName(runSpec.serviceAccount().get()))
+          .endSecret()
+          .endVolume();
+      container
+          .addToVolumeMounts(new VolumeMountBuilder()
+              .withMountPath(STYX_WORKFLOW_SA_SECRET_MOUNT_PATH)
```
Comment (fabriziodemaria) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: What happens in the unlikely case that the mounting path is the same of the one for the old-fashioned secret and both are mounted? If that not allowed, should we throw an exception with a custom message as we do, for example, in case of failure in creating the keys?

File: styx-api-service/src/test/java/com/spotify/styx/api/deprecated/WorkflowResourceTest.java (hunk 4 above)
Comment (honnix) [inline styx-api-service/src/test/java/com/spotify/styx/api/deprecated/WorkflowResourceTest.java]: Let's keeping voting for it. :D

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 7 above)
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: I wouldn't see that happening in real world though. Let's keep it in mind, and do it in another PR if really need. This PR is big enough. ;)

**Comment by danielnorberg** [comment]
🎉
